### PR TITLE
Adds Wandb + validation loss

### PR DESCRIPTION
**Pull request by RobotSail**
This PR adds support for logging through Wandb and validation loss calculation

### CODE REVIEW

File: src/mini_trainer/setup_model_for_training.py [hunk 1]
```diff
@@ -302,46 +309,89 @@ def setup_model(
...
+    patch_target_module(
+        "transformers.loss.loss_utils.fixed_cross_entropy",
+        hf_fixed_cross_entropy_none_reduction,
+    )
+    """need to patch the loss function to not reduce, so we can reduce across all GPUs"""
+    patch_target_module(
+        "liger_kernel.transformers.model.loss_utils.fixed_fused_linear_cross_entropy",
+        liger_fixed_fused_linear_cross_entropy_none_reduction,
+    )
+    ModelClass = AutoLigerKernelForCausalLM if use_liger_kernels else AutoModelForCausalLM
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:331]: When will liger not be available?

File: src/mini_trainer/async_structured_logger.py [hunk 2]
```diff
@@ -44,6 +54,13 @@ async def log(self, data):
             data["timestamp"] = datetime.now().isoformat()
             self.logs.append(data)
             await self._write_logs_to_file(data)
+            
+            # Log to wandb if enabled and wandb is initialized, but only log this on the MAIN rank
+            if self.use_wandb and wandb is not None and wandb.run is not None and dist.get_rank() == 0:
+                # Create a copy of data without timestamp for wandb (wandb handles timestamps)
+                wandb_data = {k: v for k, v in data.items() if k != "timestamp"}
+                wandb.log(wandb_data)
+
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/async_structured_logger.py]: We don't use DDP, only FSDP. So we always expect this to be set, no?

File: src/mini_trainer/setup_model_for_training.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:331]: @coderabbitai When will liger not be available?

[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:331]: @coderabbitai I think you're mistaken, liger kernel is currently a required dependency. Flash attention is the only extra dep

File: src/mini_trainer/setup_model_for_training.py [hunk 3]
```diff
@@ -7,10 +7,11 @@
 from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
     checkpoint_wrapper as ptd_checkpoint_wrapper,
 )
+import torch.nn as nn
 from torch.distributed.device_mesh import init_device_mesh
-from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig, Mxfp4Config
+from transformers import AutoTokenizer, AutoConfig, Mxfp4Config
 from mini_trainer.utils import get_model_class_from_config, log_rank_0, patch_target_module
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:13]: We bound our version of transformers so it's at least version v4.55.0, which should have the Mxfp4Config.

File: src/mini_trainer/setup_model_for_training.py [hunk 4]
```diff
@@ -302,25 +302,26 @@ def setup_model(
 
     tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
 
-    if use_liger_kernels:
-        """need to patch the loss function to not reduce, so we can reduce across all GPUs"""
-        from mini_trainer.none_reduction_losses import (
-            liger_fixed_fused_linear_cross_entropy_none_reduction,
-        )
+    # patch both loss functions, since models will use the regular HF
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/setup_model_for_training.py:305]: Why the code change here? Can't we leave as is and follow the conditional logic when liger is used.

File: src/mini_trainer/training_types.py [hunk 5]
```diff
@@ -83,3 +83,15 @@ class TrainingArgs:
...
     train_dtype: str = field(default="float32", metadata={"help": "Dtype for training computations. Can be 'float16', 'bfloat16', 'float32', etc."})
 
+    # Weights & Biases integration
+    wandb_project: Optional[str] = field(default=None, metadata={"help": "Weights & Biases project name."})
+    wandb_run_name: Optional[str] = field(default=None, metadata={"help": "Weights & Biases run name."})
+    wandb_entity: Optional[str] = field(default=None, metadata={"help": "Weights & Biases entity/team name."})
+
+    # validation
+    validation_split: float = field(default=0.0, metadata={"help": "The fraction of data to use for validation. 0.0 means no validation, 0.1 means 10% of the data is used for validation."})
+    validation_frequency: int = field(default=100, metadata={"help": "The frequency of validation in steps."})
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/training_types.py:93]: Do we want this to be 0 since default for `validation_split` is 0.0 and get the user to set it as required?

File: src/mini_trainer/sampler.py [hunk 6]
```diff
@@ -299,43 +406,93 @@ def get_data_loader(
...
+        seed=seed
+    )
+    if val_dataset is not None:
+        log_rank_0(f"Dataset split: {len(train_dataset)} train, {len(val_dataset)} validation samples")
+    else:
+        log_rank_0(f"Dataset split: {len(train_dataset)} train")
+
+    
+    # Create collate function
+    collate_fn = MaxTokensPerRankCollator(
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/sampler.py:450]: Okay nice so validation happens in a distributed fashion too using `MaxTokensPerRankCollator` class.

**Review by NikhilNayak-debug — APPROVED** [review]
Changes look good to me, just couple of small comments! Thanks for the work!

File: src/mini_trainer/setup_model_for_training.py (hunk 4 above)
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:305]: There's 2 reasons for this change:

1. Liger will always be present and isn't like flash-attention where you need to have CUDA-specific libraries installed for it. 
2. When we switch the model to eval mode, it will still use the default cross-entropy loss from transformers instead of the Liger fused cross-entropy. So we need to patch both in order for validation to also have the reduction set properly.

Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py:305]: @NikhilNayak-debug 

> we expose liger flag in training args as an option yet create this hard dependency even when user doesn't want to use it

Yeah, the current change is to patch both modules. But the choice of which model to use (liger or not) still falls down to the user, which you can see happening on this line: 

https://github.com/Red-Hat-AI-Innovation-Team/mini_trainer/pull/38/files/14ab269d78200e87278d4643cea828fc56b6621e#diff-224ccb10135078c1647c871bdc90ce96d82253b5355ba8421f9961b4ae292381R323

Currently Liger gets installed unconditionally, since it doesn't have the same build issues that flash-attention does: 
https://github.com/Red-Hat-AI-Innovation-Team/mini_trainer/blob/main/pyproject.toml#L43

This change just patches both modules so that their usage is consistent with how we expect loss to be outputted & reduced.

--- 

> Second point, wait eval should also work as expected with the corresponding patch right?

When the model is in train mode (happens during the main training loop), it uses the correct LigerFusedCrossEntropy loss module. To calculate validation loss, we switch the model into eval mode, which adjusts the model's forward pass. A consequence of this is that the model will switch back to using the standard HF CrossEntropyLoss function instead of the FusedLigerCrossEntropy loss. 

If we were to conditionally patch Liger, the correct code would look like this:

```py
    # HF cross entropy gets patched since it will always be used by either training or eval
    # therefore we patch it unconditionally
    patch_target_module(
        "transformers.loss.loss_utils.fixed_cross_entropy",
        hf_fixed_cross_entropy_none_reduction,
    )
    ModelClass = AutoModelForCausalLM  # this is already the default case
    
    # This applies the patch conditionally - since these would only be used by training if enabled
    if use_liger_kernels:
        patch_target_module(
            "liger_kernel.transformers.model.loss_utils.fixed_fused_linear_cross_entropy",
            liger_fixed_fused_linear_cross_entropy_none_reduction,
        )
        ModelClass = AutoLigerKernelForCausalLM
```

Technically either of these are valid, but since both modules will always be installed and therefore will be available for patching, it's simpler to just patch both and then use the model class which the user requested.
